uint/src/uint.rs

```diff
@@ -1550,31 +1576,34 @@ macro_rules! construct_uint {
 			}
 		}
 
-		$crate::impl_std_for_uint!($name, $n_words);
-		// `$n_words * 8` because macro expects bytes and
-		// uints use 64 bit (8 byte) words
-		$crate::impl_quickcheck_arbitrary_for_uint!($name, ($n_words * 8));
-		$crate::impl_arbitrary_for_uint!($name, ($n_words * 8));
-	}
-}
-
-#[cfg(feature = "std")]
-#[macro_export]
-#[doc(hidden)]
-macro_rules! impl_std_for_uint {
-	($name: ident, $n_words: tt) => {
 		impl $crate::core_::str::FromStr for $name {
-			type Err = $crate::rustc_hex::FromHexError;
+			type Err = $crate::FromHexError;
 
 			fn from_str(value: &str) -> $crate::core_::result::Result<$name, Self::Err> {
-				use $crate::rustc_hex::FromHex;
-				let bytes: Vec<u8> = match value.len() % 2 == 0 {
-					true => value.from_hex()?,
-					false => ("0".to_owned() + value).from_hex()?,
-				};
+				const BYTES_LEN: usize = $n_words * 8;
+				const MAX_ENCODED_LEN: usize = BYTES_LEN * 2;
+
+				let mut bytes = [0_u8; BYTES_LEN];
+
+				let encoded = value.as_bytes();
 
-				if $n_words * 8 < bytes.len() {
-					return Err(Self::Err::InvalidHexLength);
+				if encoded.len() > MAX_ENCODED_LEN {
+					return Err($crate::hex::FromHexError::InvalidStringLength.into());
+				}
+
+				if encoded.len() % 2 == 0 {
+					let out = &mut bytes[BYTES_LEN - encoded.len() / 2..];
+
+					$crate::hex::decode_to_slice(encoded, out).map_err(Self::Err::from)?;
+				} else {
+					// Prepend '0' by overlaying our value on a scratch buffer filled with '0' characters.
+					let mut s = [b'0'; MAX_ENCODED_LEN];
+					s[MAX_ENCODED_LEN - encoded.len()..].copy_from_slice(encoded);
+					let encoded = &s[MAX_ENCODED_LEN - encoded.len() - 1..];
```

Comment:
hmm, `MAX_ENCODED_LEN - encoded.len() - 1..` is used to insert a b'0' before the hex_str when the length/number of bytes is odd?

why can't we just do:

```rust
let mut s = [b'0'; MAX_ENCODED_LEN];
s[1..].copy_from_slice(encoded);
let encoded = [..encoded.len() +1]
```

Reply:
`encoded` can be of any length that is less than `MAX_ENCODED_LEN`. Your snippet will panic if `encoded`'s length is not `MAX_ENCODED_LEN - 1` but, say, `MAX_ENCODED_LEN - 3`.

Reply:
right,